sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -1155,4 +1155,54 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     val primitiveUDF = udf((i: Int) => i * 2)
     checkAnswer(df.select(primitiveUDF($"age")), Row(44) :: Row(null) :: Nil)
   }
+
+  test("SPARK-12398 truncated toString") {
+    val df1 = Seq((1: Long, "row1": String)).toDF("id", "name")
+    assert(df1.toString() === "[id: bigint, name: string]")
+
+    val df2 = Seq((1: Long, "c2": String, false: Boolean)).toDF("c1", "c2", "c3")
+    assert(df2.toString === "[c1: bigint, c2: string ... 1 more field]")
+
+    val df3 = Seq((1: Long, "c2": String, false: Boolean, 10: Integer)).toDF("c1", "c2", "c3", "c4")
+    assert(df3.toString === "[c1: bigint, c2: string ... 2 more fields]")
+
+    val df4 = Seq((1: Long, Tuple2(1: Long, "val": String))).toDF("c1", "c2")
+    assert(df4.toString === "[c1: bigint, c2: struct<_1: bigint,_2: string>]")
+
+    val df5 = Seq((1: Long, Tuple2(1: Long, "val": String), 20.0: Double)).toDF("c1", "c2", "c3")
+    assert(df5.toString === "[c1: bigint, c2: struct<_1: bigint,_2: string> ... 1 more field]")
+
+    val df6 =
+      Seq((1: Long,
+        Tuple2(1: Long, "val": String),
+        20.0: Double, 1: Integer)).toDF("c1", "c2", "c3", "c4")
+    assert(df6.toString === "[c1: bigint, c2: struct<_1: bigint,_2: string> ... 2 more fields]")
```

Comment:
Its a little odd that there are spaces after the top level commas but not the ones inside of struct.